include/ck/tensor_operation/gpu/element/element_wise_operation.hpp

```diff
@@ -389,6 +389,49 @@ struct UnaryTypeConvert<ck::bhalf_t, float>
     }
 };
 
+struct ScaleMask
+{
+    ScaleMask(float scale, float mask_filter_value)
+        : scale_(scale), mask_filter_value_(mask_filter_value)
+    {
+    }
+    // scale, masked
+    template <typename Y, typename X0, typename X1>
+    __host__ __device__ constexpr void operator()(Y& y, const X0& x, const X1& mask) const;
+
+    __host__ __device__ constexpr void
+    operator()(float& y, const float& x, const int32_t& mask) const
```

Comment:
I doubt the necessity of adding int32_t mask to ck's library instance. On my side, I find that adding int32_t is much slower that adding `F16`

```
                Ternary      785.91 us 65.84 tflops float16 B=64 S=512 T=512 N=12 H=64 mask_2d   ScaleMask
                If else      785.40 us 65.88 tflops float16 B=64 S=512 T=512 N=12 H=64 mask_2d   ScaleMask in if else
                Arithmatic   800.28 us 64.65 tflops float16 B=64 S=512 T=512 N=12 H=64 mask_2d   x * scale + (1-mask) * mask_filter_value_
                + mask (i32) 775.17 us 66.75 tflops float16 B=64 S=512 T=512 N=12 H=64 mask_2d   x * scale + mask, ignore the correctness, for perf testing
                + mask (i8)  750.57 us 68.94 tflops float16 B=64 S=512 T=512 N=12 H=64 mask_2d   x * scale + mask, ignore the correctness, for perf testing, D0sType is ck::Tuple<int8_t>, operator()(..., const int8_t& mask)
                + mask (f16) 696.89 us 74.25 tflops float16 B=64 S=512 T=512 N=12 H=64 mask_2d   x * scale + mask, ignore the correctness, for perf testing, D0sType is ck::Tuple<F16>, operator()(..., const F16& mask)
convert then    + mask (f16) 706.09 us 73.28 tflops float16 B=64 S=512 T=512 N=12 H=64 mask_2d   x * scale + mask, ensure the correctness, for perf testing, D0sType is ck::Tuple<F16>, operator()(..., const F16& mask), 
```
The `convert then    + mask` is that I use another kernel launch for materializing the `(mask == 1 ? 0.0f : mask_filter_value_)` and write it global memory buffer and then launch a `F16` `ScaleAdd` version Flash Attention

Reply:
Digging a little bit further, https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/06f1fc864c47f68fadb9c92f8f762a7bf83c15f6/include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_multiple_d_softmax_gemm_xdl_cshuffle_v1.hpp#L868-L897

To me, the gemm1 part is compute bounded, the D0s elementwise op part will be memory bounded. I think these parts don't fit quite well in this case. In gemm, you generally use software pipeline to manually hide the latency. The hardware cannot help with the memory bounded load because of low occupancy. Unless you move the D0s load before gemm1, I think there is no easy fix for this :(